concurrency/DCMetricsUtils.java

```diff
@@ -0,0 +1,481 @@
+    private static String methodName;
+    private static String teamName;
+    private static String siteVariant;
+
+    private boolean verbose;
+
+    private DCMetricsUtils(boolean verboseLogging) {
+        verbose = verboseLogging;
+    }
+
+    /**
+     * @return instance of DCMetricsUtils
+     */
+    public static DCMetricsUtils getInstance() {
+        return getInstance(false);
+    }
+
+    /**
+     * Fetches the singleton instance.
+     *
+     * @param verboseLogging True if each metric sent should be logged.
+     * @return instance of DCMetricsUtils
+     */
+    public static DCMetricsUtils getInstance(boolean verboseLogging) {
+        if (!isInitialized) {
+            Slog.e(TAG, "Unable to fetch an instance since the DCM library is not initialized");
+            throw new IllegalStateException("DCM not properly initialized");
+        }
+
+        if (dcMetricsUtils == null) {
+            dcMetricsUtils = new DCMetricsUtils(verboseLogging);
+        }
+
+        return dcMetricsUtils;
+    }
+
+    /**
+     * Initializes the DCM library to start being used.
+     * @param newConfig The configuration that you want to have for the DCM settings
+     */
+    public static void initialize(Context context, DCMetricsConfig newConfig) {
+        dcmConfig = newConfig;
+        serviceName = dcmConfig.getServiceName();
+        methodName = dcmConfig.getMethodName();
+        teamName = dcmConfig.getTeamName();
+        siteVariant = dcmConfig.getSiteVariant();
+        mFactory = AndroidMetricsFactoryImpl.getInstance(context);
+        AndroidMetricsFactoryImpl.setOAuthHelper(context, dcmConfig.getOAuthHelper());
+        AndroidMetricsFactoryImpl.setDeviceType(context, dcmConfig.getDeviceType());
+        AndroidMetricsFactoryImpl.setDeviceId(context, dcmConfig.getDeviceId());
+        isInitialized = true;
+    }
+
+    /**
+     * @return Whether or not the DCM Utilities have been properly initialized
+     */
+    public static boolean isInitialized() {
+        return isInitialized;
+    }
+
+    /**
+     * Disables the DCM library from usage and marks it as no longer initialized
+     */
+    public static void disable() {
+        isInitialized = false;
+    }
+
+    /**
+     * Add a MetricEvent object to the METRIC_INDEX using the metricId as a key.  If a MetricEvent already exists using the metricId, nothing is done.
+     */
+    public DCMetricsUtils createMetric(String metricId) {
+        return createMetric(metricId, MetricEventType.getDefault());
+    }
+
+    public DCMetricsUtils createMetric(String metricId, MetricEventType eventType) {
+        Slog.v(TAG, "Create Metric: " + metricId);
+
+        if (!METRIC_INDEX.containsKey(metricId)) {
+            MetricEvent metricEvent = mFactory.createMetricEvent(serviceName, methodName, eventType);
+            metricEvent.addString(DCMetricsConfig.SESSION_ID_LABEL, dcmConfig.getSessionId());
+            METRIC_INDEX.put(metricId, metricEvent);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.



**Problem**  
 You are using a `ConcurrentHashMap`, but your usage of `containsKey()` and `put()` may not be thread-safe at lines: **114 and 117**. Two threads can perform this **same check at the same time** and one thread can **overwrite** the value written by the other thread.

**Fix**  
 Consider replacing `put()` with `putIfAbsent()` to help prevent accidental overwriting. `putIfAbsent()` puts the value only if the `ConcurrentHashMap` does not contain the key and therefore avoids overwriting the value written there by the other thread&#39;s `putIfAbsent()`.

**More info**  
 `putIfAbsent()` returns `null` if the value did not exist and returns the value in the map if one already exists.

 [View an example on GitHub](https://github.com/apache/hadoop/blob/f16cf877e565084c66bc63605659b157c4394dc8/hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/hdfs/server/namenode/startupprogress/StartupProgress.java#L272-L273) (external link).